src/providers/GoodsSearchProvider.php

```diff
@@ -0,0 +1,409 @@
+            'properties' => $properties
+        ];
+    }
+
+    /**
+     * @return Sort
+     */
+    private function getSort()
+    {
+        $attributes = [];
+
+        foreach ($this->sortFields as $sortField) {
+            $attributes[$sortField] = [
+                'asc' => [$sortField => SORT_ASC],
+                'desc' => [$sortField => SORT_DESC],
+            ];
+        }
+
+        $sort = new Sort(
+            [
+                'attributes' => $attributes,
+                'sortParam' => $this->sortParameter
+            ]
+        );
+
+        return $sort;
+    }
+
+    private function getGoodsIds($query)
+    {
+        return $query->column();
+    }
+
+    /**
+     * @param $fullSectionArray
+     * @param $sections
+     * @return mixed
+     */
+    private function getTree($fullSectionArray, &$sections)
+    {
+        foreach ($sections as $key => $section) {
+            if (array_key_exists($section['id'], $fullSectionArray)) {
+                $sections[$key]['subsections'] = $fullSectionArray[$section['id']];
+                $this->getTree($fullSectionArray, $sections[$key]['subsections']);
+            }
+        }
+        return $sections;
+    }
+
+    /**
+     * @param $goodsIds
+     * @return mixed
+     */
+    private function getGoodsSectionsTree($goodsIds)
+    {
+        $categoryGoodsModels = CategoryGoods::find()
+            ->select('structure_id')
+            ->addSelect('goods_id')
+            ->where(
+                [
+                    'goods_id' => $goodsIds
+                ]
+            )
+            ->all();
+
+        $categoryGoods = [];
+        foreach ($categoryGoodsModels as $section) {
+            $categoryGoods[$section->structure_id][] = $section->goods_id;
+        }
+
+        $goodsCategoryModels = BaseStructure::find()
+            ->innerJoin(
+                GoodsCategoryExtended::tableName(),
+                GoodsCategoryExtended::tableName() . '.[[model_id]] = ' . BaseStructure::tableName() . '.[[id]]'
+            )
+            ->where(
+                [
+                    'is_active' => 1,
+                    'is_deleted' => 0
+                ]
+            )
+            ->all();
+
+        $goodsCategory = [];
+
+        foreach ($goodsCategoryModels as $section) {
+            $goodsCategory[$section->parent_id][] = [
+                'name' => $section->defaultTranslation->name,
+                'id' => $section->id,
+                'parentId' => $section->parent_id,
+                'slug' => $section->defaultTranslation->slug,
+                'count' => isset($categoryGoods[$section->id]) ? count($categoryGoods[$section->id]) : 0
+            ];
+        }
+
+        $firstCategory = reset($goodsCategory);
+
+        return $this->getTree($goodsCategory, $firstCategory);
+    }
+
+    public function getEntities(&$actionData)
+    {
+        if (($searchPhrase = \Yii::$app->request->get($this->searchParameter)) !== null) {
+
```

Comment:
add a comment `// @todo: rewrite this code via search component (yii2-data-structure-tools module)`